modin/test/storage_formats/pandas/test_internals.py

```diff
@@ -209,3 +248,109 @@ def row_apply_func(row):
     ), "Partitions are not block partitioned after apply."
+
+
+@pytest.mark.skipif(
+    Engine.get() not in ("Dask", "Ray"),
+    reason="Only Dask and Ray engines have virtual partitions.",
+)
+@pytest.mark.parametrize(
+    "axis,virtual_partition_class",
+    ((0, virtual_column_partition_class), (1, virtual_row_partition_class)),
+    ids=["partitions_spanning_all_columns", "partitions_spanning_all_rows"],
+)
+class TestBuildVirtualPartition:
+    def test_from_virtual_partitions_with_call_queues(
+        self, axis, virtual_partition_class
+    ):
+        # reverse the dataframe along the virtual partition axis.
+        def reverse(df):
+            return df.iloc[::-1, :] if axis == 0 else df.iloc[:, ::-1]
+
+        level_zero_blocks_first = [
+            block_partition_class(put(pandas.DataFrame([0]))),
+            block_partition_class(put(pandas.DataFrame([1]))),
+        ]
+        level_one_virtual_first = virtual_partition_class(
+            level_zero_blocks_first, full_axis=False
+        )
+        level_one_virtual_first = level_one_virtual_first.add_to_apply_calls(reverse)
+        level_zero_blocks_second = [
+            block_partition_class(put(pandas.DataFrame([2]))),
+            block_partition_class(put(pandas.DataFrame([3]))),
+        ]
+        level_one_virtual_second = virtual_partition_class(
+            level_zero_blocks_second, full_axis=False
+        )
+        level_one_virtual_second = level_one_virtual_second.add_to_apply_calls(reverse)
+        level_two_virtual = virtual_partition_class(
+            [level_one_virtual_first, level_one_virtual_second], full_axis=True
+        )
+        level_two_virtual_result = level_two_virtual.apply(lambda df: df, num_splits=1)[
+            0
+        ]
+        if axis == 0:
+            expected_df = pandas.DataFrame([1, 0, 3, 2], index=[0, 0, 0, 0])
+        else:
+            expected_df = pandas.DataFrame([[1, 0, 3, 2]], columns=[0, 0, 0, 0])
+        df_equals(
+            level_two_virtual_result.to_pandas(),
+            expected_df,
+        )
+
+    def test_from_block_and_virtual_partition_with_call_queues(
+        self,
+        axis,
+        virtual_partition_class,
+    ):
+        # make a function that reverses the dataframe along the virtual
+        # partition axis.
+        # for testing axis == 0, start with two 2-rows-by-1-column blocks. for
+        # axis == 1, start with two 1-rows-by-2-column blocks.
+        def reverse(df):
+            return df.iloc[::-1, :] if axis == 0 else df.iloc[:, ::-1]
+
+        block_data = [[0, 1], [2, 3]] if axis == 0 else [[[0, 1]], [[2, 3]]]
+        level_zero_blocks = [
+            block_partition_class(put(pandas.DataFrame(block_data[0]))),
+            block_partition_class(put(pandas.DataFrame(block_data[1]))),
+        ]
+        level_zero_blocks[0] = level_zero_blocks[0].add_to_apply_calls(reverse)
+        level_one_virtual = virtual_partition_class(level_zero_blocks[1])
```

Comment:
Don't we need full axis set to False here? Or should we be making an axis partition along the orthogonal axis below?

Reply:
You bring up a good point; I wasn't thinking much about `full_axis`. I made it a parameter to a test class so we test full and non-full for both axis cases. I also made this test explicitly about `drain_call_queue` and stopped calling `apply`.